types/sharedb/lib/sharedb.d.ts

```diff
@@ -142,19 +128,46 @@ export class Doc<T = any> extends EventEmitter {
     flush(): void;
 }
 
-export type QueryEvent = 'ready' | 'error' | 'changed' | 'insert' | 'move' | 'remove' | 'extra';
-export class Query extends EventEmitter {
+export interface DocEventMap<T> {
+    'load': () => any;
+    'no write pending': () => any;
+    'nothing pending': () => any;
+    'create': (source: any) => any;
+    'op': (ops: any[], source: any) => any;
+    'op batch': (ops: any[], source: any) => any;
+    'before op': (ops: any[], source: any) => any;
+    'before op batch': (ops: any[], source: any) => any;
+    'del': (data: T, source: any) => any;
+    'error': (error: Error) => any;
+    'destroy': () => any;
```

Comment:
We could have all of these use `=> void` since that still accepts functions that return values.

For example, this works:
```ts
const foo: (() => void) = () => 12;
```